### PR TITLE
feat(lsp): suppress-rule, show-documentation CodeActions and iterative fix-all

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

Adapts three patterns from the [ESLint VSCode extension](https://github.com/microsoft/vscode-eslint) to tally's LSP server:

- **Suppress Rule CodeAction** (`quickfix.suppress.tally`): generates "Suppress {rule} for this line/file" actions that inject `# tally ignore=` directives. Uses cached parse result + `directive.Parse()` for existing directive detection. Smart merging, comment-block-aware insertion (avoids triggering `buildkit/InvalidDefinitionDescription`), respects `RequireReason`. Carries stable `Data` field for future "suppress and report false positive" composition.

- **Show Documentation CodeAction**: command-based action that sends `window/showDocument` to open rule docs in the system browser. Gated on client `ShowDocumentClientCapabilities.Support`. New `tally.openRuleDoc` server command.

- **Iterative Fix-All**: `computeFixEdits` now supports `"all"` mode (iterative lint+fix up to 10 passes until convergence) and `"problems"` mode (single-pass, previous behavior). Default is `"all"`.

### Supporting changes

- Expanded `lintCacheEntry` to retain `*config.Config` and `*dockerfile.ParseResult` alongside violations
- Three new settings (`suppressRuleEnabled`, `showDocumentationEnabled`, `fixAllMode`) wired through Go server, VS Code extension, and IntelliJ plugin

## Test plan

- [x] Unit tests for suppress (insertion, merging, indentation, RequireReason, dedup, comment-block-aware)
- [x] Unit tests for show-doc (DocURL present/absent, dedup, out-of-range, arg parsing)
- [x] Unit tests for settings (defaults, explicit values, invalid fixAllMode fallback)
- [x] LSP integration test snapshots updated
- [x] `make test` — 6183 tests pass
- [x] `make lint` — clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * "Fix all mode" setting (choices: "all" or "problems") controlling fix-all behavior and UI in IntelliJ/VS Code.
  * Code action to suppress rule violations by inserting per-line or file-level ignore directives.
  * Code action to open rule documentation in the browser.
  * New editor settings to enable/disable suppress and documentation actions.

* **Tests / Reliability**
  * Added tests and snapshots covering fix-all behavior, suppress actions, and documentation actions.

### CODE REVIEW

File: internal/lspserver/suppress.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+		// Found an existing directive — check if it already suppresses this rule.
+		if d.SuppressesRule(v.RuleCode) {
+			return nil // already suppressed
+		}
+		// Merge: append the rule code to the existing directive line.
+		return mergeRuleIntoDirectiveLine(d.Line, v.RuleCode, lines)
+	}
+
+	// No existing directive — insert a new one above the comment block.
+	insertLine0 := findCommentBlockStart(violationLine0, lines)
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lspserver/suppress.go]: ✅ Addressed in 4a2b182. Thanks @gemini-code-assist!

`findCommentBlockStart` now stops at parser directives (`# syntax=`, `# escape=`, `# check=`) and treats bare `#` lines as block separators. Added regression tests for both edge cases.

File: internal/lspserver/suppress.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+// directive should be inserted — after any # syntax= or # escape= parser
+// directives and their trailing blank lines.
+func findInsertionAfterParserDirectives(lines []string) int {
+	line := 0
+	for line < len(lines) {
+		trimmed := strings.TrimSpace(lines[line])
+		lower := strings.ToLower(trimmed)
+		if strings.HasPrefix(lower, "# syntax=") ||
+			strings.HasPrefix(lower, "# escape=") ||
+			strings.HasPrefix(lower, "# check=") {
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lspserver/suppress.go]: ✅ Addressed in 79ed055. Thanks @gemini-code-assist!

Replaced `isParserDirective()` and `findInsertionAfterParserDirectives()` with `firstInstructionLine()` which derives the boundary from the already-parsed AST (`parseResult.AST.AST.Children[0].StartLine`). No string-based parser directive detection at all — the AST is the authoritative source.

File: internal/lspserver/fixall.go [hunk 3]
```diff
@@ -77,7 +77,52 @@ func fixModeAllowsSafeFix(ruleCode string, fixModes map[string]fix.FixMode) bool
...
+	for range maxFixIterations {
+		if ctx.Err() != nil {
+			return nil
+		}
+		next := s.computeFixedContent(ctx, docURI, current, safety)
+		if next == nil || bytes.Equal(next, current) {
+			break
+		}
+		current = next
+	}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lspserver/fixall.go:129]: ✅ Addressed in bd1c90e. Thanks @gemini-code-assist!

Config resolution and fix mode computation are now hoisted out of the iteration loop. `computeFixedContent` is a standalone function that receives pre-resolved `cfg` and `fixModes` — no more repeated `resolveConfig` disk I/O or `fix.BuildFixModes` on each pass.

File: internal/lsptest/__snapshots__/TestLSP_ExecuteCommandApplyAllFixesRealWorld_1.snap.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+	&& echo "${OPEN_MPI_PATH}/bin/mpirun.real --allow-run-as-root \"\$@\"" >> $OPEN_MPI_PATH/bin/mpirun \
+	&& chmod a+x $OPEN_MPI_PATH/bin/mpirun \
+	&& echo "hwloc_base_binding_policy = none" >> $OPEN_MPI_PATH/etc/openmpi-mca-params.conf \
+	&& echo "rmaps_base_mapping_policy = slot" >> $OPEN_MPI_PATH/etc/openmpi-mca-params.conf \
+	&& echo NCCL_DEBUG=INFO >> /etc/nccl.conf \
+	&& echo NCCL_SOCKET_IFNAME=^docker0 >> /etc/nccl.conf
+RUN mkdir /tmp/efa-ofi-nccl && cd /tmp/efa-ofi-nccl && git clone https://github.com/aws/aws-ofi-nccl.git -b v${BRANCH_OFI} && cd aws-ofi-nccl && ./autogen.sh && ./configure --with-libfabric=/opt/amazon/efa --with-mpi=/opt/amazon/openmpi --with-cuda=/usr/local/cuda --with-nccl=/usr/local --prefix=/usr/local && make && make install && rm -rf /tmp/efa-ofi-nccl && rm -rf /var/lib/apt/lists/* && apt-get clean
+RUN apt-get update \
+	&& apt-get install -y --allow-downgrades --allow-change-held-packages --no-install-recommends \
+	&& apt-get install -y --no-install-recommends openssh-client openssh-server \
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lsptest/__snapshots__/TestLSP_ExecuteCommandApplyAllFixesRealWorld_1.snap.Dockerfile:304]: Acknowledged — as noted in the analysis, this malformed split is pre-existing in the formatting snapshot and not introduced by this PR. The iterative fix-all simply inherits the same output from the underlying fixer. Filed as a separate concern for the Dockerfile fixer.

File: internal/lspserver/suppress.go [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+		vRange := violationRange(v)
+		if !rangesOverlap(vRange, params.Range) {
+			continue
+		}
+
+		key := seenKey{ruleCode: v.RuleCode, line: v.Location.Start.Line}
+		if _, ok := seen[key]; ok {
+			continue
+		}
+		seen[key] = struct{}{}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lspserver/suppress.go:63]: ✅ Addressed in 6223eca. Thanks @coderabbitai!

Added a separate `seenFile` map keyed by `ruleCode` only for file-level dedup. Per-line actions still use `(ruleCode, line)`. Regression test confirms same rule on two lines produces 2 line actions + 1 file action.

File: _integrations/intellij-tally/src/main/kotlin/io/github/wharflab/tally/intellij/lsp/TallySettings.kt [hunk 6]
```diff
@@ -29,6 +32,9 @@ internal object TallySettings {
             fixUnsafe = service.fixUnsafe,
             configurationOverride = service.configurationPath?.takeIf { it.isNotBlank() },
             workspaceTrusted = workspaceTrusted,
+            suppressRuleEnabled = true,
+            showDocumentationEnabled = true,
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply _integrations/intellij-tally/src/main/kotlin/io/github/wharflab/tally/intellij/lsp/TallySettings.kt]: ✅ Addressed in cec2e92. Thanks @coderabbitai!

Added `suppressRuleEnabled` and `showDocumentationEnabled` to `TallySettingsState` (default `true`) with accessors in `TallySettingsService`, and wired them through `fromService()` instead of hard-coding.

File: internal/lspserver/suppress.go [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+		}
+		if !d.AppliesTo.Contains(violationLine0) {
+			continue
+		}
+		// Found an existing directive — check if it already suppresses this rule.
+		if d.SuppressesRule(v.RuleCode) {
+			return nil // already suppressed
+		}
+		// Merge: append the rule code to the existing directive line.
+		return mergeRuleIntoDirectiveLine(d.Line, v.RuleCode, lines)
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lspserver/suppress.go]: ✅ Addressed in 1819b86. Thanks @coderabbitai!

Now only merges into `SourceTally` directives. Still checks `SuppressesRule` across all sources (tally, hadolint, buildx) to avoid inserting a redundant directive when a foreign directive already suppresses the rule.

File: internal/lsptest/__snapshots__/TestLSP_CodeAction_1.snap.json [hunk 8]
```diff
@@ -67,5 +67,97 @@
...
+        "character": 0,
+        "line": 0
+       }
+      }
+     }
+    ]
+   }
+  },
+  "kind": "quickfix.suppress.tally",
+  "title": "Suppress tally/newline-between-instructions for this file"
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lsptest/__snapshots__/TestLSP_CodeAction_1.snap.json:177]: ✅ Addressed in a5f682d. Thanks @coderabbitai!

The data was being set correctly by the server, but the test's `codeAction` struct was missing a `Data` field — JSON deserialization silently dropped it. Added the field and updated the snapshot. All four suppress actions now show `{type, ruleCode}` data.

File: internal/lspserver/suppress.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lspserver/suppress.go]: ✅ Addressed in e2cf132. Thanks @gemini-code-assist!

`suppressLineEdit` now resolves the violation line to the instruction's start line via `InstructionSpanIndex.ContainingSpan()`. Added `TestSuppressLineEdit_HeredocViolation` — a `RUN <<EOF` heredoc with a ShellCheck violation on line 4 correctly produces a directive above line 2 (the RUN instruction), not inside the heredoc body.

File: internal/lspserver/fixall.go [hunk 9]
```diff
@@ -77,8 +77,75 @@ func fixModeAllowsSafeFix(ruleCode string, fixModes map[string]fix.FixMode) bool
...
+	fixModes := fix.BuildFixModes(cfg)
+
+	current := content
+	for range maxFixIterations {
+		if ctx.Err() != nil {
+			return nil
+		}
+		next := computeFixedContent(ctx, filePath, current, cfg, fixModes, safety)
+		if next == nil || bytes.Equal(next, current) {
+			break
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lspserver/fixall.go]: ✅ Addressed in e851142. Thanks @coderabbitai!

`computeFixedContent` now returns `([]byte, error)`. The iterative loop checks the error separately — on failure it returns `nil` (no edits) instead of returning partially-fixed content. `(nil, nil)` = converged, `(nil, err)` = failed.

File: internal/lspserver/suppress.go [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+	}
+
+	sm := sourcemap.New(parseResult.Source)
+	spanIndex := directive.NewInstructionSpanIndexFromAST(parseResult.AST, sm)
+	dirResult := directive.Parse(sm, nil, spanIndex)
+
+	// The first instruction's start line (0-based) marks where parser directives
+	// end. Derived from the already-parsed AST — no string-based parser directive
+	// detection needed.
+	firstInstLine0 := firstInstructionLine(parseResult)
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lspserver/suppress.go]: ✅ Addressed in 4b3c986. Thanks @coderabbitai!

Replaced `firstInstructionLine()` with `parserDirectiveFloor()` which uses the first AST child's `PrevComment` length to compute the line after the last parser directive. Now `findCommentBlockStart` can walk through description comments above the first instruction, placing the suppress directive above them (not between description and FROM).

Added `TestSuppressLineEdit_DescriptionAboveFirstInstruction` — verifies that with `# syntax=` + description comment + FROM, the directive goes above the description comment.

File: internal/lspserver/suppress.go [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+	insertPos := len(line)
+	if idx := strings.Index(line, ";reason="); idx >= 0 {
+		insertPos = idx
+	}
+
+	// Trim trailing whitespace before insertion point.
+	trimmed := insertPos
+	for trimmed > 0 && line[trimmed-1] == ' ' {
+		trimmed--
+	}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lspserver/suppress.go]: ✅ Addressed in 96f4429. Thanks @gemini-code-assist!

`AppendRule` now trims both spaces and tabs before the insertion point. Added test case for tab trimming.

**Comment by tinovyatkin** [comment]
@codex review

File: internal/lspserver/suppress.go [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+		// Next-line suppress action
+		if edit := suppressLineEdit(v, lines, dirResult, requireReason); edit != nil {
+			suppressLineData := any(map[string]any{"type": "suppress-line", "ruleCode": v.RuleCode})
+			actions = append(actions, protocol.CodeAction{
+				Title: fmt.Sprintf("Suppress %s for this line", v.RuleCode),
+				Kind:  ptrTo(suppressCodeActionKind),
+				Edit: &protocol.WorkspaceEdit{
+					Changes: new(map[protocol.DocumentUri][]*protocol.TextEdit{
+						uri: {edit},
+					}),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lspserver/suppress.go:76]: This is a false positive — `new(composite_literal)` compiles and works correctly in Go. The same pattern is used throughout the existing codebase (`codeactions.go:166`, `execute_command.go:52`, `fixall.go:30-31`). Go's `new` builtin with a composite literal evaluates to a pointer to the literal value.

[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/lspserver/suppress.go:76]: Correction to my earlier reply: `new(value)` is a Go 1.26 feature (not standard in older Go). It returns `*T` pointing to the value — a shorthand for `&value` that works with unaddressable expressions like composite literals.

Added 584381a with assertions that `Edit.Changes` is non-nil, contains edits for the document URI, and the edits have non-empty `NewText` — directly verifying the `new(map[...]{...})` produces correct `*map` values.

File: internal/directive/format.go [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+// AppendRule computes the edit needed to append ruleCode to an existing
+// directive line (e.g. "# tally ignore=DL3008" → insert ",DL3027").
+//
+// The edit inserts before ";reason=" if present, otherwise at end of line,
+// trimming trailing whitespace before the insertion point.
+func AppendRule(lineText, ruleCode string) AppendRuleEdit {
+	// Find insertion point: before ;reason= if present, otherwise at end.
+	insertPos := len(lineText)
+	if idx := strings.Index(lineText, reasonSeparator); idx >= 0 {
+		insertPos = idx
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/directive/format.go]: ✅ Addressed in 70cd9c3. Thanks @chatgpt-codex-connector!

`AppendRule` now uses a regex (`(?i);reason\s*=\s*`) matching the parser's syntax instead of exact `";reason="` string. Handles `;reason = ...`, `;REASON=...`, and other valid variants. Added test cases for both.
